### PR TITLE
[SYCL-MLIR][sycl-raise-host] Handle `parallel_for` "lambda" arguments allocated in a calling function

### PR DESCRIPTION
**Pull request by victor-eds**
Handle "lambda" arguments of `parallel_for` allocated in calling function, i.e., the case in which `parallel_for` is not inlined.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 1]
```diff
@@ -1680,6 +1682,61 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
...
+    auto parentFunction = dyn_cast<LLVM::LLVMFuncOp>(block->getParentOp());
+    if (!parentFunction)
+      return nullptr;
+
+    // Search function argument
+    const auto *blockArg = llvm::find(block->getArguments(), lambdaObj);
+    if (blockArg == block->args_end())
+      return nullptr;
+
+    // Recurisvely search for alloca defining operation.
```
Comment (jopperm) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:1711]:
```suggestion
    // Recursively search for alloca defining operation.
```

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 2]
```diff
@@ -1680,6 +1682,61 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
...
+  /// Return the `llvm.alloca` operation defining \p lambdaObj.
+  static LLVM::AllocaOp getAlloca(Value lambdaObj) {
+    if (Operation *definingOp = lambdaObj.getDefiningOp()) {
+      // lambdaObj defined in this function
+      return TypeSwitch<Operation *, LLVM::AllocaOp>(definingOp)
+          .Case<LLVM::AllocaOp>([](auto alloca) { // Return the defining alloca
+            return alloca.getElemType().has_value() ? alloca : nullptr;
+          })
+          // Safely bail out
+          .Default(LLVM::AllocaOp());
```
Comment (jopperm) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:1694]:
```suggestion
          .Default(nullptr);
```
Nit: elsewhere you use `nullptr` to signal failure.

**Review by jopperm — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp (hunk 2 above)
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp]: Slightly changed so that it compiles

**Comment by jopperm** [comment]
Should we KISS the raising for now and close or delay this PR in favor of #10665?

**Comment by victor-eds** [comment]
> Should we KISS the raising for now and close or delay this PR in favor of #10665?

I'll close it. Will retake if needed in the future.

**Pull request closed without merging**